### ISSUE TITLE
trying to stop CI running two copies of each test, but gcovr is now failing

### ISSUE THREAD

**Issue by jbathegit** [NOAA-EMC/NCEPLIBS-bufr#240]
Hi @edwardhartnett, I'm trying to implement your suggestion from [grib_util 165](https://github.com/NOAA-EMC/NCEPLIBS-grib_util/issues/165) into the bufr library.

In the bufr library we have two different yaml files, one (main.yml) for all of the regular builds on all of the varying OS and compiler combinations, and a second one (main_new.yml) for running gcovr with the GNU v10 compiler on ubuntu.  It's this second one that is now failing.

For the record, here are the changes I made to these files, according to your suggestion in the above grib_util issue:

![yaml diffs](https://user-images.githubusercontent.com/62722702/202747726-5c5f1b3b-63bc-4128-8dba-c5fc828ab46e.png)

With these changes, everything still works fine in main.yml, but main_new.yml is now generating the errors shown in https://github.com/NOAA-EMC/NCEPLIBS-bufr/actions/runs/3492827265/jobs/5847112538  If you scroll up in that output, you can see that all of the normal tests are still passing in the ctest step, but then an "Assertion error" gets generated when trying to run gcovr.

Any idea what may be going on here?  I'm also CCing @AlexanderRichert-NOAA in case he may have any suggestions.

**Comment by jbathegit** [comment]
I forgot to mention that the suggested changes do have the intended effect of now only running the CI for pull_request actions and not also for push actions, so that's working as expected, and thanks for that tip!

However, as noted above, gcovr no longer seems to be a happy camper.

**Comment by jbathegit** [comment]
Hi @edwardhartnett and @AlexanderRichert-NOAA, just touching base to see if you have any ideas on what may be going on here?  As noted, I was trying to follow the advice suggested in [grib_util 165](https://github.com/NOAA-EMC/NCEPLIBS-grib_util/issues/165), but gcovr no longer works and is now throwing an "Assertion error" after making the suggested changes to main_new.yml.

Also tagging @aerorahul and @Hang-Lei-NOAA in case they may have seen this sort of thing before with other libraries.

**Comment by jbathegit** [comment]
@edwardhartnett @AlexanderRichert-NOAA @aerorahul @Hang-Lei-NOAA

I tried reverting main_new.yml back to what it was originally, but gcovr still isn't working (see https://github.com/NOAA-EMC/NCEPLIBS-bufr/actions/runs/3605654044/jobs/6076325402 (from the pull request) and/or https://github.com/NOAA-EMC/NCEPLIBS-bufr/actions/runs/3605653952/jobs/6076302902 (from the push)).

So the problem is something other than the changes I had previously made to main_new.yml.  Has anyone else seen anything similar when running gcovr for other libraries or GitHub projects?

**Comment by AlexanderRichert-NOAA** [comment]
@jbathegit I'm not too versed with gcovr... It looks like it could be a gcovr bug (has the version changed recently?), or it appears this can happen if there are two functions with the same name.

**Comment by jbathegit** [comment]
Thanks for the reply @AlexanderRichert-NOAA.  Sadly, I don't know much about gcovr either.  @edwardhartnett set this up several years ago, and the yaml file has been untouched since then.

I also have no idea what version of gcovr is being used by the GitHub runner, or how to even find that out.  According to the [gcovr GitHub repository](https://github.com/gcovr/gcovr), the latest release version is 5.2, but I don't know how to confirm if that's what the runner is using.  The job output doesn't print what version it's using.

I've been doing a deep dive myself, searching for any guidance on the Internet.  I also saw something where it mentioned that gcovr could get hung up if there were two files with the same name, but that shouldn't be the case here.  The only thing is that the test codes have recently been renamed as *.F files, because they now have Fortran preprocessor code in them.  So maybe gcovr is getting hung up by that?

**Comment by AlexanderRichert-NOAA** [comment]
@jbathegit How tricky would it be to download those gcovr-related files? I think this could be done by adding a CI step that makes them artifacts even when the job fails, something like:
```
- name: Archive gcovr files
  if: always()
  uses: actions/upload-artifact@v1
  with:
    name: gcovr-files
    path: path/to/files
```

**Comment by jbathegit** [comment]
Thanks, but the bigger problem is that the code coverage check is currently a "must pass" test.  In other words, we can't merge any new PRs into the main `develop` branch for this library until we figure this out.  Or unless we remove that yaml file altogether from the repository, which I'm not personally advocating for since I think the gcovr output is useful.

**Comment by AlexanderRichert-NOAA** [comment]
@jbathegit Looks like it's a problem with gcovr. I was able to reproduce the problem on my personal machine when using gcovr 5.1 or 5.2 installed through pip (pip is installing 5.2 in these failed runs, which you can find if you search for "gcovr" in the logs). Anyway, I was able to make it work using 5.0, so for now I would recommend changing "sudo python3 -m pip install -U gcovr" to "sudo python3 -m pip install -U gcovr**==5.0**", then file a bug with gcovr and see if they think it's a bona fide bug.

**Comment by edwardhartnett** [comment]
Turn off the gcovr and clear out your backlog of merges. The code coverage is meant to help, not slow things down.

I will take a look at the CI system this week. I have been bringing all the other projects up to better standard of testing...

**Comment by jbathegit** [comment]
Dang, I just merged all of @edwardhartnett's recent updates from #243 back into our long-running `r8wraps` branch.  That branch is nearly finished to resolve #78, but when I tried rerunning my test PR #239 on that updated branch, that same gcovr error popped up again in the developer.yml test.

All of the other .yml tests passed fine, but you can see the same error again in https://github.com/NOAA-EMC/NCEPLIBS-bufr/actions/runs/3652803297/jobs/6171594904 if you search for the string "Trouble" in the output from the test step.

@edwardhartnett I can try the above suggested fix by @AlexanderRichert-NOAA to force the use of gcovr v5.0, but I was really hoping your recent updates would have resolved this issue as well.  The `r8wraps` branch does have a lot of commits in it, but the only thing that's changed in the file test/test_OUT_3.F (which is where gcovr is now squawking) is to add some preprocessor lines to the test code so that we can test it compiling with 8-byte integers.  So in turn, I had to rename the file from its original test/test_OUT_3.f (lowercase f) to test/test_OUT_3.F (uppercase F) so that the preprocessor would run on it.  So I really don't understand why gcovr is now squawking about this.

When I was getting this same error earlier (see above comments) it was in different test/test*F codes, but it's always been one of those codes that gcovr ends up choking on.  I never had any problems with any of these test programs until I added in the preprocessor changes and renamed them with the uppercase F.  You can see exactly what was changed in [this diff output](https://github.com/NOAA-EMC/NCEPLIBS-bufr/pull/239/commits/9cd22f54b16f205b0e7d871e016d8a10b2cdf66f#diff-77e894c408c269695c03df79833ce60576cbe5fd1710020c0547ed3338f5febe) and just scroll down to the test codes.

Again, I can try forcing the use of gcovr 5.0, but it seems like I'm missing something basic here.

**Comment by jbathegit** [comment]
Actually, and come to think of it, why is gcovr even running on the test/test* codes?  Those test codes are designed to exercise the various parts of the library, and gcovr is running on all of the library routines, but why does it also need to run on the test codes themselves?

I notice the gcovr command now contains the option `--excludes ../tests`, so maybe you were already thinking the same thing, and if that's the case then maybe it's just a typo, because those codes are all in the `../test` directory, not `../tests`!

**Comment by edwardhartnett** [comment]
Indeed gcovr should not run on test codes.

The test coverage is running well in the new develop branch - I just checked it. And we will not try to do code coverage in any other workflow.

So I think this issue can be closed.

**Comment by jbathegit** [comment]
Fixing the typo in the option to `--excludes ../test` did the trick, and now it's working!

Thanks everyone for your help on this!

**Comment by edwardhartnett** [comment]
OK, it seems clear that the gcovr problem had nothing to do with the changes to how often the CI runs, but just happened at the same time because the utility was updated or changed in some way.

Indeed I did set up the original system some time ago and have no real memory of the details of course. But the latest and greatest way of running gcovr is now happening in the developer workflow and should not happen anywhere else, as it is just as waste of cycles (we only need one).
